src/StorageFolding.cpp

```diff
@@ -196,29 +239,49 @@ class IsBufferSpecial : public IRVisitor {
 
 // Look for opportunities for storage folding in a statement
 class StorageFolding : public IRMutator {
+    const map<string, Function> &env;
+
     using IRMutator::visit;
 
     void visit(const Realize *op) {
         Stmt body = mutate(op->body);
 
-        AttemptStorageFoldingOfFunction folder(op->name);
         IsBufferSpecial special(op->name);
         op->accept(&special);
 
+        // Do explicit storage folding first.
+        auto func_it = env.find(op->name);
+        Function func = func_it != env.end() ? func_it->second : Function();
+
         if (special.special) {
+            for (const StorageDim &i : func.schedule().storage_dims()) {
+                user_assert(!i.fold_factor.defined())
+                    << "Dimension " << i.var << " of " << op->name
+                    << " cannot be folded because it is accessed by extern or device stages.\n";
+            }
+
             debug(3) << "Not attempting to fold " << op->name << " because its buffer is used\n";
             if (body.same_as(op->body)) {
                 stmt = op;
             } else {
                 stmt = Realize::make(op->name, op->types, op->bounds, op->condition, body);
             }
         } else {
+            AttemptStorageFoldingOfFunction folder(func);
             debug(3) << "Attempting to fold " << op->name << "\n";
-            Stmt new_body = folder.mutate(body);
+            if (folder.dims_folded.empty()) {
```

Comment:
I don't get this. What if there's a totally unrelated Func that needs folding inside the body?